### ISSUE TITLE
The proxy get riak_repl2_pg_block_provider.erl timer handling looks too odd [JIRA: RIAK-1807]

### ISSUE THREAD

**Issue by slfritchie** [basho/riak_repl#679]
See the end of this commentary for suggestions for changing the timer handling code in riak_repl2_pg_block_provider.erl

---

From: https://basho.zendesk.com/agent/tickets/10882

Wow, something is really wrong with that https://basho.zendesk.com/agent/#/tickets/10793 cluster dump. That is 8.3 million 'keepalive' messages. If I used some Perl to replace all of the 'keepalive,' strings to make it easier to read, then I see only this:

Message queue: [KEEPALIVE-REPEATS,{tcp,#Port&lt;0.340692>,&lt;&lt;174 bytes>>},KEEPALIVE-REPEATS,keepalive]
So, there's only one non-keepalive message out of over 8.3 million.

If the timer service sends one message/second, then that is over 96 days of keepalive messages. If the timer process was correct and sends only 1 msg/sec, and if the process was over 96 days old, then the gen_server was stuck for over 96 days and there was only 1 PDU of TCP data in the same 96 days. That scenario seems .... very unlikely?

If the OS system time went backward for 96 days, then perhaps the timer service had a "catch up" of sending 8M messages very quickly? I don't know, that sounds .... very unlikely also?

HRM, this code looks weird, it should be using erlang:send_interval() or erlang:start_timer() or perhaps timer:send_after() as a one-time thing, rather than as a repeated timer.

https://github.com/basho/riak_repl/blob/develop/src/riak_repl2_pg_block_provider.erl#L163-L164

The handle_info() callbacks cancels the repeated timer and then re-created here:

https://github.com/basho/riak_repl/blob/develop/src/riak_repl2_pg_block_provider.erl#L131-L134

If keepalive_timer() is changed to use a 1-time timer, then there are some places where new calls to keepalive_timer() must be added, e.g. https://github.com/basho/riak_repl/blob/develop/src/riak_repl2_pg_block_provider.erl#L109-L112

**Comment by jonmeredith** [comment]
Confirm it is very odd indeed. There's a bug in canceling the request timers - the state holds the return value from timer:send_interval rather than the reference, so the timer:cancel() call errors (without being checked).

```
{state,ranch_tcp,#Port<0.340692>,undefined,"remote",#Ref<0.0.1.236681>,undefined,{riak_client,'nodename',undefined},{ok,{interval,#Ref&lt;0.0.3714.150360>}},10806}
```

Recommended fix
- Replace timer:send_interval / timer:cancel call with `erlang:send_after` and `erlang:cancel_timer` and make sure timers are rescheduled correctly for all cases.
- Maybe pattern match on the return value for canceling timers so we'll know if we make this mistake again.
- Drop the frequency from 1s to 60s. It makes no difference to the protocol as the far side just consumes the message and does nothing.

@shino I know I mentioned this to you in chat today, if you did anything on it please comment here, otherwise we'll let you know when a patch is available.

This issue has been present since 2013/02/05 when the code was first written and will only affect users using RiakCS+MDC.

**Comment by Basho-JIRA** [comment]
pr: https://github.com/basho/riak_repl/pull/680 

**_[posted via JIRA by Micah Warren]**_

**Comment by Basho-JIRA** [comment]
Jon M confirmed code merged to 2.0.6 on 6/25/15.  Dev is done. Now just waiting to go into additional releases.  

**_[posted via JIRA by Patricia Brewer]**_
